Add option to write icon to directory - fixes #3

SamVerschueren opened the pull request:
Gave a shot at implementing #3. Feedback time :)!

sindresorhus commented:
The point of this, at least from my perspective, was to increase performance by making Swift write the file directly instead of all the overhead of passing it through stdout and then writing it with Node.js.

SamVerschueren commented:
Good point! Will have a look at it. Never done something with Swift so might be a nice excercise.

sindresorhus commented:
Happy to guide you in the right direction. You could use `CLI.args[2]` for the destination and if it's supplied it means the `dest` option (Let's make that `dest` => `destination`) is set. 

I think you could do something like this on line 77 (untested):

```swift
if (CLI.args[2]) {
	do {
		try iconData.write(to: URL(fileURLWithPath: CLI.args[2]), options: .atomic)
	} catch {
		CLI.printErr(error)
		exit(1)
	}

	return
}
```

SamVerschueren commented:
Tried your code, had to change one small thing but works fine!

The only thing I did though in the current JavaScript implementation is use `filenamify` to create the filename from the input app/path/...

So currently, maybe we should just discuss the API first as well. Especially if we still want to implement #2.

I think we have 2 options here. The first one is add 1 uniform interface for 1 icon and for multiple icons. The second one, and maybe my preferred way of doing it is adding an extra method.

```js
fileIcon('Safari', {destination: __dirname + '/Safari.png'}).then(buffer => { });

fileIcon.all([
    {
        application: 'Safari',
        size: 512,
        destination: path.join(__dirname, 'Safari_512.png')
    },
    {
        application: 'Safari',
        size: 1024,
        destination: path.join(__dirname, 'Safari_1024.png')
    }
]).then(() => {
    // Doesn't return anything
    console.log('done');
});
```

sindresorhus commented:
Let's go with the second one (`.all()`). `application` => `app`.

SamVerschueren commented:
So destination will have the entire path right? Including the filename?

sindresorhus commented:
Yes

SamVerschueren commented:
This should work. Not sure about the docs yet though.

sindresorhus commented:
If `destination` is specified, I don't think we should return a Buffer, for performance reasons. Maybe they should really be separate methods too? `fileIcon.file()`, `fileIcon.buffer()`, `fileIcon.multiple()`?

SamVerschueren commented:
I don't think that we actually return something when `destination` is set because in Swift we [exit early](https://github.com/sindresorhus/file-icon/pull/4/files#diff-9dd60505f1f18051b9c5dceff9eb0fd7R83). But yes, maybe we should. That would actually make more sense.

SamVerschueren commented on readme.md:
```diff
@@ -42,14 +46,41 @@ Either:
+
+### options
+
+Type: `object`
```
Should we just state this instead?

> Same `options` as `fileIcon.buffer`

sindresorhus replied:
They don't have the same options. `fileIcon.file()` also has `destination`.

SamVerschueren replied:
Yes that line + the destination option. But I'm fine if you are with the current docs :)

SamVerschueren commented:
I added `fileIcon.buffer` and `fileIcon.file`. Not sure about the API though. I moved `size` to an options object as that one is always optional. But what about destination? Is it okay to have it in the `options` object or should we add it as second parameter.

```js
fileIcon.buffer('Safari', {size: 64}).then(() => { });

fileIcon.file('Safari', 'safari-icon.png', {size: 64}).then(() => { });
```

sindresorhus commented on index.js:
```diff
@@ -25,15 +26,34 @@ module.exports = (file, opts) => {
+	return execa.stdout(bin, [file, opts.size], {encoding: 'buffer'});
+};
+
+exports.file = (file, opts) => {
```
I've started a new pattern. By wrapping the whole thing in a promise I can be absolute sure nothing throws synchronously and I don't have to have all the verbose `return Promise.reject()` stuff.

`exports.file = (file, opts) => {` => `exports.file = (file, opts) => Promise.resolve().then({`

When we support async functions, we can simply remove the wrapper and it will work fine.

SamVerschueren replied:
Nice idea!

sindresorhus commented on readme.md:
```diff
@@ -16,23 +16,23 @@ $ npm install --save file-icon
 
-fileIcon('Safari', {destination: 'safari-icon.png'}).then(() => {
+fileIcon.file('Safari', {destination: 'safari-icon.png'}).then(() => {
 	const buffer = fs.readFileSync('safari-icon.png');
```
We already show how to get the buffer with `.buffer()`. Would make more sense to just show a `console.log('Done')` here or something.

SamVerschueren commented on index.js:
```diff
@@ -1,25 +1,48 @@
 const bin = path.join(__dirname, '.build/release/file-icon');
 
-module.exports = (file, size) => {
+const validate = (file, opts) => {
```
I moved all the validation steps to a separate function so I can reuse them which makes the exported functions a lot smaller.

sindresorhus commented:
Looks good. I did some minor tweaks.

SamVerschueren commented:
LGTM!
